src/explorer/connectionTreeItem.ts

```diff
@@ -53,58 +60,65 @@ export default class ConnectionTreeItem extends vscode.TreeItem implements TreeI
-        return Promise.resolve(this._childrenCache);
-      } else {
-        return new Promise(async (resolve, reject) => {
-          // If we aren't the active connection, we reconnect.
-          if (this._connectionController.getActiveConnectionInstanceId() !== this._connectionInstanceId) {
-            try {
-              await this._connectionController.connectWithInstanceId(this._connectionInstanceId);
-            } catch (err) {
-              return reject(err);
-            }
+    if (!this.isExpanded || this._connectionController.isDisconnecting() || this._connectionController.isConnnecting()) {
+      return Promise.resolve([]);
+    }
+
+    if (this._childrenCacheIsUpToDate) {
+      return Promise.resolve(Object.values(this._childrenCache));
+    } else {
+      return new Promise(async (resolve, reject) => {
+        const dataService = this._connectionController.getActiveConnection();
+        dataService.listDatabases((err: any, databases: string[]) => {
+          if (err) {
+            return reject(`Unable to list databases: ${err}`);
           }
 
-          const dataService = this._connectionController.getActiveConnection();
-          dataService.listDatabases((err: any, databases: string[]) => {
-            if (err) {
-              return reject(`Unable to list databases: ${err}`);
-            }
+          this._childrenCacheIsUpToDate = true;
 
-            this._childrenCacheIsUpToDate = true;
+          if (databases) {
+            const pastChildrenCache = this._childrenCache;
+            this._childrenCache = {};
 
-            if (databases) {
-              this._childrenCache = databases.map(
-                ({ name }: any) => new DatabaseTreeItem(name, dataService)
-              );
-            } else {
-              this._childrenCache = [];
-            }
+            databases.forEach(({ name }: any) => {
+              this._childrenCache[name] = pastChildrenCache[name]
+                ? new DatabaseTreeItem(name, dataService, pastChildrenCache[name].isExpanded, pastChildrenCache[name].getChildrenCache())
+                : new DatabaseTreeItem(name, dataService, false /* Collapsed */, {});
+            });
+          } else {
+            this._childrenCache = {};
+          }
 
-            return resolve(this._childrenCache);
-          });
+          return resolve(Object.values(this._childrenCache));
         });
-      }
+      });
     }
-
-    // Here we either want to return loading or nothing.
-    return Promise.resolve([]);
   }
 
   onDidCollapse() {
     this.isExpanded = false;
     this._childrenCacheIsUpToDate = false;
   }
 
-  async onDidExpand() {
+  onDidExpand(): Promise<any> {
     this._childrenCacheIsUpToDate = false;
     this.isExpanded = true;
+
+    // If we aren't the active connection, we reconnect.
+    if (this._connectionController.getActiveConnectionInstanceId() !== this._connectionInstanceId) {
+      return new Promise((resolve, reject) => {
```

Comment:
You don't need `reject` here if you use only resolve both in case of success and in case of error.